<launch>
  <!-- Debug Info -->
  <arg name="use_dummy" default="false" />
  <arg unless="$(arg use_dummy)" name="marvin_controller_args" default="" />
  <arg     if="$(arg use_dummy)" name="marvin_controller_args" default="--dummy MX-106" />

  <!-- Debug Info -->
  <arg name="debug" default="false" />
  <arg unless="$(arg debug)" name="launch_prefix" default="" />
  <arg     if="$(arg debug)" name="launch_prefix" default="gdb --ex run --args" />

  <!-- Start the Dynamixel motor manager to control all marvin servos  -->
  <group ns="marvin" >
    <node name="marvin_controller_manager" pkg="marvin_controller"
          type="controller_manager.py" required="true" output="screen"
          args="$(arg marvin_controller_args)" launch-prefix="$(arg launch_prefix)">
      <rosparam file="$(find marvin_controller)/config/dynamixel_ports.yaml" command="load"/>
    </node>

    <!-- Load the joint controller configuration from a YAML file -->
    <rosparam file="$(find marvin_controller)/config/marvin_controller_configuration.yaml"
              command="load"/>

    <!-- Start the Steering and Wheel controllers -->
    <node name="dynamixel_controller_spawner_MX106"
          pkg="dynamixel_controllers"
          type="controller_spawner.py"
          args="--manager=marvin_controller_manager
                --port=dynamixel_MX106
                --type=simple
                fl_rotation_joint_position_controller
                bl_rotation_joint_position_controller
                br_rotation_joint_position_controller
                fr_rotation_joint_position_controller
                fl_wheel_joint_velocity_controller
                bl_wheel_joint_velocity_controller
                br_wheel_joint_velocity_controller
                fr_wheel_joint_velocity_controller
                "
          output="screen" />

    <!-- Start the joint state publishers -->
    <node name="joint_state_publisher_spawner"
          pkg="dynamixel_controllers"
          type="controller_spawner.py"
          args="--manager=marvin_controller_manager
                --port=dynamixel_MX106
                --type=meta
                joint_state_publisher
                fl_rotation_joint_position_controller
                bl_rotation_joint_position_controller
                br_rotation_joint_position_controller
                fr_rotation_joint_position_controller
                fl_wheel_joint_velocity_controller
                bl_wheel_joint_velocity_controller
                br_wheel_joint_velocity_controller
                fr_wheel_joint_velocity_controller
                "
          output="screen" />
  </group>
</launch>
